src/libraries/System.Private.CoreLib/generators/EventSourceGenerator.Emitter.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Text;
+using System.Threading;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Generators
+{
+    public partial class EventSourceGenerator
+    {
+        private class Emitter
+        {
+            private StringBuilder _builder = new StringBuilder(1024);
+            private GeneratorExecutionContext _context;
+
+            public Emitter(GeneratorExecutionContext context) => _context = context;
+
+            public void Emit(EventSourceClass[] eventsources, CancellationToken cancellationToken)
+            {
+                foreach (var ec in eventsources)
+                {
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        // stop any additional work
+                        break;
+                    }
+
+                    _builder.AppendLine("using System;");
+                    GenType(ec);
+
+                    _context.AddSource($"{ec.ClassName}.Generated", SourceText.From(_builder.ToString(), Encoding.UTF8));
+
+                    _builder.Clear();
+                }
+            }
+
+            private void GenType(EventSourceClass ec)
+            {
+                if (!string.IsNullOrWhiteSpace(ec.Namespace))
+                {
+                    _builder.AppendLine($@"
+namespace {ec.Namespace}
+{{");
+                }
+
+                _builder.AppendLine($@"
+    partial class {ec.ClassName}
+    {{");
+                GenerateConstructor(ec);
+
+                GenerateProviderMetadata(ec.SourceName);
+
+                _builder.AppendLine($@"
+    }}");
+
+                if (!string.IsNullOrWhiteSpace(ec.Namespace))
+                {
+                    _builder.AppendLine($@"
+}}");
+                }
+            }
+
+            private void GenerateConstructor(EventSourceClass ec)
+            {
+                _builder.AppendLine($@"
+        private {ec.ClassName}() : base(new Guid({ec.Guid.ToString("x").Replace("{", "").Replace("}", "")}), ""{ec.SourceName}"") {{ }}");
+            }
+
+            private void GenerateProviderMetadata(string sourceName)
+            {
+                _builder.Append(@"
+        private protected override ReadOnlySpan<byte> ProviderMetadata => new byte[] { ");
+
+                byte[] metadataBytes = MetadataForString(sourceName);
+                foreach (var b in metadataBytes)
+                {
+                    _builder.Append($"0x{b:x}, ");
```

Comment:
How many bytes does this typically end up writing out in this fashion?  I'm curious what this does to compilation times.

Reply:
The runtime eventsources go via the basic path with no `m_traits` so is quite short e.g. `PortableThreadPoolEventSource` is

![image](https://user-images.githubusercontent.com/1142958/104762309-0a849180-575c-11eb-99b3-122759419b3c.png)